In [50]:
# preparing environment

import pandas as pd
import numpy as np
import scipy.stats as stats
import matplotlib.pyplot as plt
import seaborn as sns


from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import export_graphviz
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import f1_score

In [51]:
# imports
from env import host, user, password
import pandas as pd
import numpy as np
import os

# create url to connect to database
def get_connection(db, user=user, host=host, password=password):
    return f'mysql+pymysql://{user}:{password}@{host}/{db}'
    
# pass sql query to database retrieve data
def new_zillow_data():
    sql_query = "SELECT * FROM properties_2017 JOIN predictions_2017 on predictions_2017.parcelid = properties_2017.parcelid WHERE unitcnt = 1 AND transactiondate BETWEEN '2017-05-01' AND '2017-06-30'"
    df = pd.read_sql(sql_query, get_connection('zillow'))
    df.to_csv('zillow_df.csv')
    return df

In [52]:
df = new_zillow_data()

## PREPARE

Prepare, tidy, and clean the data so it can be explored and analyzed.
***

In [30]:
# displaying column names and null value counts
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13309 entries, 0 to 13308
Data columns (total 63 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   id                            13309 non-null  int64  
 1   parcelid                      13309 non-null  int64  
 2   airconditioningtypeid         6226 non-null   float64
 3   architecturalstyletypeid      0 non-null      object 
 4   basementsqft                  0 non-null      object 
 5   bathroomcnt                   13309 non-null  float64
 6   bedroomcnt                    13309 non-null  float64
 7   buildingclasstypeid           2 non-null      float64
 8   buildingqualitytypeid         13162 non-null  float64
 9   calculatedbathnbr             13296 non-null  float64
 10  decktypeid                    0 non-null      object 
 11  finishedfloor1squarefeet      0 non-null      object 
 12  calculatedfinishedsquarefeet  13308 non-null  float64
 13  f

Dropping columns that meet any of the following criterion:
* more than 2,000 null values 
* do not contain relevant data for predicting property value 
    - examples include irrelevant IDs, fips, etc.
* taxamount, structuretaxvaluedollarcnt, landtaxvaluedollarcnt 
    - per project specifications we cannot use these columns
        - "for prediction purposes, we won't know tax amount until we know tax value"
        - confirmed by Data Science senior team member John Salas

In [34]:
# .nunique shows us how many unique values are held in each column
df.nunique()

id                              13308
parcelid                        13308
bathroomcnt                        12
bedroomcnt                         12
buildingqualitytypeid              12
calculatedfinishedsquarefeet     3079
fips                                2
heatingorsystemtypeid               3
latitude                        12063
longitude                       11258
lotsizesquarefeet                8058
propertycountylandusecode          31
propertylandusetypeid               9
propertyzoningdesc               1212
rawcensustractandblock          10183
regionidcity                      123
regionidcounty                      2
regionidzip                       277
roomcnt                             1
unitcnt                             1
yearbuilt                         124
structuretaxvaluedollarcnt      10329
taxvaluedollarcnt               11074
assessmentyear                      1
landtaxvaluedollarcnt           10091
taxamount                       13212
censustracta

Dropping columns that meet any of the following criterion:
- categorical columns with more than 5 unique values 
    - it will be cumbersome to create a boolean value for each
- columns with only 1 value as 
    - our model won't be able to make any meaningful distinctions with them

While viewing data in SQL we noted several columns that appear to have identical data to one another. We will confirm if this is the case.

In [55]:
# comparing each pair of columns we suspected are duplicates
# the resulting variables tell us how many rows differ between each
sqft_columns_diff = (df.finishedsquarefeet12 != df.calculatedfinishedsquarefeet).sum()
bathroom_count_diff = (df.calculatedbathnbr != df.bathroomcnt).sum()
bathroom_count_diff_alt = (df.fullbathcnt != df.bathroomcnt).sum()

print(f'Number of different values between finishedsquarefeet12 and calculatedfinishedsquarefeet: {sqft_columns_diff}')
print(f'Number of different values between calculatedbathnbr and bathroomcnt: {bathroom_count_diff}')
print(f'Number of different values between fullbathcnt and bathroomcnt: {bathroom_count_diff_alt}')

Number of different values between finishedsquarefeet12 and calculatedfinishedsquarefeet: 13
Number of different values between calculatedbathnbr and bathroomcnt: 13
Number of different values between fullbathcnt and bathroomcnt: 13


Dropping the following columns because with the exception of 13 rows, they are duplicates:
- finishedsquarefeet12 (near-duplicate of calculatedfinishedsquarefeet)
- calculatedbathnbr (near-duplicate of bathroomcnt)
- bathroomcnt (near-duplicate of bathroomcnt)

In [57]:
# Dropping columns with more than 2000 rows missing
df = df.drop(columns=['airconditioningtypeid', 'architecturalstyletypeid', 'basementsqft','yardbuildingsqft26', 'buildingclasstypeid', 'decktypeid', 'finishedfloor1squarefeet', 'finishedsquarefeet13', 'finishedsquarefeet15', 'finishedsquarefeet50', 'finishedsquarefeet6', 'fireplacecnt', 'garagecarcnt', 'garagetotalsqft', 'hashottuborspa', 'poolcnt', 'poolsizesum', 'pooltypeid10', 'pooltypeid2', 'pooltypeid7', 'regionidneighborhood', 'storytypeid', 'threequarterbathnbr', 'typeconstructiontypeid', 'yardbuildingsqft17', 'numberofstories', 'fireplaceflag', 'taxdelinquencyflag', 'taxdelinquencyyear'], axis=1)

# Dropping columns with high amount of duplicate values
df = df.drop(columns=['finishedsquarefeet12', 'calculatedbathnbr', 'fullbathcnt'], axis=1)

# Dropping columns with irrelevant data
df = df.drop(columns=['id', 'fips', 'parcelid', 'latitude', 'longitude', 'rawcensustractandblock', 'censustractandblock', 'heatingorsystemtypeid', 'regionidcounty', 'logerror' ], axis=1)

# Dropping categorical columns with more than 5 unique values
df = df.drop(columns=['propertycountylandusecode', 'propertylandusetypeid','buildingqualitytypeid', 'propertyzoningdesc', 'regionidcity', 'regionidzip', 'yearbuilt', 'transactiondate' ], axis=1)

# Dropping columns with only 1 unique value
df = df.drop(columns=['roomcnt', 'unitcnt', 'assessmentyear'], axis=1)

# Dropping columns that are prohibited from use
df = df.drop(columns=['taxamount', 'structuretaxvaluedollarcnt', 'landtaxvaluedollarcnt'], axis=1)

All columns that we specified for removal have been dropped. We'll now take another look at the data with .info for further changes needed.

In [58]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13309 entries, 0 to 13308
Data columns (total 5 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   bathroomcnt                   13309 non-null  float64
 1   bedroomcnt                    13309 non-null  float64
 2   calculatedfinishedsquarefeet  13308 non-null  float64
 3   lotsizesquarefeet             12976 non-null  float64
 4   taxvaluedollarcnt             13309 non-null  float64
dtypes: float64(5)
memory usage: 520.0 KB


We will change the names of our remaining columns to make them easier to understand
- Rename bathroomcnt to bathroom_count
- Rename bedroomcnt to bedroom_count
- Rename calculatedfinishedsquarefeet to unit_sq_feet
- Rename lotsizesquarefeet to lot_size_sq_feet
- Rename taxvaluedollarcnt to tax_dollar_value

lotsizesquarefeet has 333 missing values
- Out of 13309 rows, this is less than 3% of the total so we'll remove them along with the one null value row from calculatedfinishedsquarefeet

In [60]:
# using head to examine raw values of columns
df.head(10)

bathroomcnt  bedroomcnt  calculatedfinishedsquarefeet  lotsizesquarefeet  \
0          2.0         3.0                        1458.0             8284.0   
1          1.0         2.0                        1421.0             6707.0   
2          2.0         3.0                        1650.0             7300.0   
3          1.0         2.0                         693.0             2908.0   
4          1.0         1.0                         812.0            93226.0   
5          2.0         3.0                        1108.0             6050.0   
6          2.0         3.0                        1947.0             6404.0   
7          1.0         2.0                        1061.0             4800.0   
8          2.0         3.0                        1632.0             6040.0   
9          3.0         3.0                        2768.0            34309.0   

   taxvaluedollarcnt  
0           136104.0  
1            35606.0  
2           614000.0  
3           274237.0  
4           245906.0  
5           486866.0  
6           388478.0  
7            31839.0  
8           499402.0  
9           151340.0

- No binary values
- Various units of measurement (dollars, feet, bedrooms, bathrooms)
- All columns contain numerical values

For these reasons we will need to scale all of our columns except taxvaluedollarcnt (since it's our target variable and target variables should not be scaled).

## Prepare Takeaways

Removing columns that meet any of the following criterion:
- More than 2,000 row missing
- Duplicates or near-duplicates of other columns
- Contain data that's irrelevant to our project
- Categorical with more than 5 unique values
- Contain only 1 unique column
- Are prohibited from being used per project spefications

Rename remaining columns for readability and understanding

Drop rows with missing values since this accounts for less than 3% of all rows

Scale every non-target variable column. (ie. everything except taxvaluedollarcnt)

Split all data into train, validate and test samples

All of these changes will be done via various functions within the prepare.py file deliverable